source/tutorials/towards-reproducability-pinning-nixpkgs.rst

```diff
@@ -0,0 +1,72 @@
+Towards reproducability: Pinning nixpkgs
+========================================
+
+In Nix snippets around the internet you'll often notice the following:
+
+.. code:: nix
+
+    { pkgs ? import <nixpkgs> {}
+    }:
+
+    ...
+
+To quickly demonstrate and get a working Nix expression by importing Nix packages.
+
+But it doesn't make Nix expression reproducible. Two developers on different machines
+are likely to get different result.
+
+.. note::
+
+  ``<nixpkgs>`` is syntax for looking up from shell environment variable ``$NIX_PATH``. 
+  
+  It is always set at the installation time to point to ``nixpkgs-unstable`` channel. 
+  
+  Channels are a way of distributing Nix software, but they are being phased out.
+  So even though they are still used by default, it's recommended to avoid channels 
+  and ``<nixpkgs>`` by always setting ``NIX_PATH=`` to be empty.
+
+
+Pinning with URLs inside Nix expression
+---------------------------------------
+
+The simplest way to pin nixpkgs is to fetch them as a tarball specified via git commit:
+
+.. code:: nix
+
+    { pkgs ? import (fetchTarball https://github.com/NixOS/nixpkgs/archive/3590f02e7d5760e52072c1a729ee2250b5560746.tar.gz) {};
+    }:
+
+    ...
+
+Picking the commit is easiest done via `status.nixos.org <https://status.nixos.org/>`_,
+which lists all the releases and their latest commit that passed all the tests.
+
+It's recommended to either follow latest stable NixOS release such as ``nixos-20.03``
+or unstable via ``nixos-unstable``.
+
+
+Dependency management with niv
+------------------------------
+
+If you'd like a bit more automation around bumping dependencies such as nixpkgs,
+``niv`` is made for exactly that::
+
+    $ niv init
+
+This command will generate ``nix/sources.json`` with information how and where
+dependencies are fetched and ``nix/sources.nix`` that glues them together in Nix.
+
+By default ``niv`` will configure the latest stable NixOS release.
+
+You can use it as:
+
+.. code:: nix
+
+    { sources ? import ./sources.nix 
+    , pkgs ? import sources.nixpkgs {}
+    }:   
+
+    ...
+
+
```

Comment:
Would be great to also show how to (periodically) update to latest all-tests-passing commit in the chosen NixOS release branch. Bonus karma points to show how to setup a GitHub Action that opens a PR to bump the commit on the first of every month.

Reply:
I feel it's better to ship a template repository since that's more hands on, rather than teaching that in a tutorial. Either way, it would have to be a separate tutorial.

Opened https://github.com/domenkozar/nix.dev/issues/25

Reply:
Fair point. That said, the tutorial still doesn't tell me how do I update `sources.json` when/if I want to. I assume it's a single `niv` command?

Reply:
Good point - done!